### PR TITLE
require_verification_before_update

### PR DESCRIPTION
**Pull request by lgallard**
Change @endrec implementation for a simpler definition of `attributes_require_verification_before_update`

Take into account `attributes_require_verification_before_update` must be the same as `auto_verified_attributes`, but instead of validating in code we can let Terraform validate it for us when running a plan / apply.

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -321,5 +321,5 @@ locals {
 
   # user_attribute_update_settings
   # As default, all auto_verified_attributes will become attributes_require_verification_before_update
-  user_attribute_update_settings = var.user_attribute_update_settings == null ? (length(var.auto_verified_attributes) > 0 ? { attributes_require_verification_before_update = var.auto_verified_attributes } : {}) : var.user_attribute_update_settings
+  user_attribute_update_settings = length(var.attributes_require_verification_before_update) > 0 ? { attributes_require_verification_before_update = var.attributes_require_verification_before_update } : {}
```
Comment (endrec) [inline main.tf:324]: This is not the same logic as it was before. This new version does not care about the `auto_verified_attributes` setting at all. 

I understand you would like to simplify the settings, so taking that, and the default into account, it should be like (the code is not tested)
```suggestion
  user_attribute_update_settings = var.attributes_require_verification_before_update == null ?
    (length(var.auto_verified_attributes) > 0 ? { attributes_require_verification_before_update = var.auto_verified_attributes } : {}) :
    ( length(var.attributes_require_verification_before_update) > 0 ? { attributes_require_verification_before_update = var.attributes_require_verification_before_update } : {} )
```

Given the security implications, I would rather require extra verification, and explicit opt-out here, so we err on the safe side.

File: variables.tf [hunk 2]
```diff
@@ -361,10 +361,10 @@ variable "tags" {
...
 
 # user_attribute_update_settings
-variable "user_attribute_update_settings" {
-  description = "Configuration block for user attribute update settings"
-  type        = map(any)
-  default     = null
+variable "attributes_require_verification_before_update" {
+  description = "A list of attributes requiring verification before update. If set, the provided value(s) must also be set in `auto_verified_attributes`. Valid values: `email`, `phone_number`"
+  type        = list(string)
+  default     = []
```
Comment (endrec) [inline variables.tf:367]: This should be null, to make the default calculation above work. And then the user have to specify an empty list for explicit opt-out.

**Comment by endrec** [comment]
> Take into account `attributes_require_verification_before_update` must be the same as `auto_verified_attributes`, but instead of validating in code we can let Terraform validate it for us when running a plan / apply.

`attributes_require_verification_before_update` must be a _subset_ of `auto_verified_attributes`, not necessarily the same. 

I was thinking about this simplification, but there must be a reason it is not this simple in the AWS terraform provider...
I suspect they will add more settings to the `user_attribute_update_settings` map later.

**Comment by lgallard** [comment]
@endrec I need to rethink this. I would like to have an easy variable input probably with a map instead of having several inputs

Renaming this PR to `require_verification_before_update`

**Comment by lgallard** [comment]
@endrec closing for now. I will implement this in another PR in a future release

**Pull request closed without merging**